plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java

```diff
@@ -310,6 +328,51 @@ private JdbcTableHandle flushAttributesAsQuery(ConnectorSession session, JdbcTab
         return Optional.of(new LimitApplicationResult<>(handle, jdbcClient.isLimitGuaranteed(session)));
     }
 
+    @Override
+    public Optional<TopNApplicationResult<ConnectorTableHandle>> applyTopN(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            long topNCount,
+            List<SortItem> sortItems,
+            Map<String, ColumnHandle> assignments)
+    {
+        if (!isTopNPushdownEnabled(session)) {
+            return Optional.empty();
+        }
+
+        verify(!sortItems.isEmpty(), "sortItems are empty");
+        for (SortItem sortItem : sortItems) {
+            verify(assignments.containsKey(sortItem.getName()), "assignments does not contain sortItem: %s", sortItem.getName());
+        }
+
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+
+        if (handle.getLimit().isPresent() && handle.getLimit().getAsLong() <= topNCount) {
+           return Optional.of(new TopNApplicationResult<>(handle, true));
+        }
+
+        List<SortItem> sortOrder = sortItems.stream()
+                .map(sortItem -> new SortItem(((JdbcColumnHandle) assignments.get(sortItem.getName())).getColumnName(), sortItem.getSortOrder()))
```

Comment:
Move the verification `verify(assignments.containsKey(sortItem.getName())` from above to here. It will be more "at use".